File: typescript/src/tools/dashboards/update.ts
import { getProjectBaseUrl } from "@/lib/utils/api";
import { DashboardUpdateSchema } from "@/schema/tool-inputs";
import { getToolDefinition } from "@/tools/toolDefinitions";
import type { Context, Tool } from "@/tools/types";
import type { z } from "zod";

const schema = DashboardUpdateSchema;

type Params = z.infer<typeof schema>;

export const updateHandler = async (context: Context, params: Params) => {
	const { dashboardId, data } = params;
	const projectId = await context.getProjectId();
	const dashboardResult = await context.api
		.dashboards({ projectId })
		.update({ dashboardId, data });

	if (!dashboardResult.success) {
		throw new Error(`Failed to update dashboard: ${dashboardResult.error.message}`);
	}

	const dashboardWithUrl = {
		...dashboardResult.data,
		url: `${getProjectBaseUrl(projectId)}/dashboard/${dashboardResult.data.id}`,
	};

	return { content: [{ type: "text", text: JSON.stringify(dashboardWithUrl) }] };
};

const tool = (): Tool<typeof schema> => ({
	name: "dashboard-update",
	description: getToolDefinition("dashboard-update").description,
	schema,
	handler: updateHandler,
});

export default tool;

File: python/tools/dashboards/update.py
import json

from api.client import is_error, is_success
from lib.utils.api import get_project_base_url
from schema.tool_inputs import DashboardUpdateSchema
from tools.tool_definitions import get_tool_definition
from tools.types import Context, Tool, ToolResult


async def update_dashboard_handler(context: Context, params: DashboardUpdateSchema) -> ToolResult:
    project_id = await context.get_project_id()
    dashboard_result = await context.api.dashboards(project_id).update(int(params.dashboardId), params.data)

    if is_error(dashboard_result):
        raise Exception(f"Failed to update dashboard: {dashboard_result.error}")

    assert is_success(dashboard_result)

    dashboard_data = dashboard_result.data

    dashboard_with_url = {
        **dashboard_data.model_dump(),
        "url": f"{get_project_base_url(project_id, context.api.base_url)}/dashboard/{dashboard_data.id}",
    }

    return ToolResult(json.dumps(dashboard_with_url))


def update_dashboard_tool() -> Tool[DashboardUpdateSchema]:
    definition = get_tool_definition("dashboard-update")

    return Tool(
        name="dashboard-update",
        description=definition["description"],
        schema=DashboardUpdateSchema,
        handler=update_dashboard_handler,
    )
